R/credentials_gce.R

```diff
@@ -74,36 +74,60 @@
 #' }
 credentials_gce <- function(scopes = "https://www.googleapis.com/auth/cloud-platform",
                             service_account = "default", ...) {
-  gargle_debug("trying {.fun credentials_gce}")
+  gargle_debug("Trying {.fun credentials_gce} ...")
   if (!is_gce()) {
+    gargle_debug(c("x" = "We don't seem to be on GCE."))
     return(NULL)
   }
 
-  gce_token <- fetch_gce_access_token(scopes, service_account = service_account)
+  requested_scopes <- normalize_scopes(scopes)
+  dat <- gce_instance_service_accounts()
+  service_account_details <- as.list(dat[dat$name == service_account, ])
+
+  account_scopes <- service_account_details$scopes
+  account_scopes <- normalize_scopes(strsplit(account_scopes, split = ",")[[1]])
+  missing <- setdiff(requested_scopes, account_scopes)
+  if (length(missing) > 0) {
+    gargle_debug(c(
+      "!" = "{cli::qty(length(missing))}{?This/These} requested \\
+             scope{?s} {?is/are} not among the scopes for the \\
+             {.val {service_account}} service account:",
+      bulletize(missing, bullet = "x"),
+      "i" = "If there are problems downstream, this might be the root cause."
+    ))
+  }
+
+  token <- gce_access_token(scopes, service_account = service_account)
 
-  params <- list(
-    as_header = TRUE,
-    scope = scopes,
-    service_account = service_account
-  )
-  token <- GceToken$new(
-    credentials = gce_token$access_token,
-    params = params,
-    cache_path = FALSE,
-    # The underlying Token2 class appears to *require* an endpoint and an app,
-    # though it doesn't use them for anything in this case.
-    endpoint = httr::oauth_endpoints("google"),
-    app = httr::oauth_app("google", key = "KEY", secret = "SECRET")
-  )
-  token$refresh()
   if (is.null(token$credentials$access_token) ||
     !nzchar(token$credentials$access_token)) {
     NULL
   } else {
+    gargle_debug("GCE service account email: {.email {service_account_details$email}}")
+    gargle_debug("GCE service account name: {.val {token$params$service_account}}")
+    gargle_debug("GCE access token scopes: {.val {commapse(base_scope(token$params$scope))}}")
     token
   }
 }
 
+#' Fetch access token for a service account on GCE
+#'
+#' @inheritParams credentials_gce
+#'
+#' @keywords internal
+#' @export
+gce_access_token <- function(scopes = "https://www.googleapis.com/auth/cloud-platform",
```

Comment:
This sort of adapter function is common in gargle's other flows. It sits between the `credential_*()` function that is suitable for inclusion in the credential registry used by `token_fetch()` and the R6 class (`GceToken`, in this case).